src/objects/set.rs

```diff
@@ -0,0 +1,269 @@
+use conversion::ToPyObject;
+use objects::PyObject;
+use err::{self, PyResult, PyErr};
+use std::{mem, collections, hash, cmp, ptr};
+
+/// Represents a Python `set`.
+pub struct PySet(PyObject);
+
+pyobject_newtype!(PySet, PySet_Check, PySet_Type);
+
+impl PySet {
+    /// Creates a new set from any iterable
+    ///
+    /// Corresponds to `set(iterable)` in Python.
+    pub fn new<I>(py: Python, iterable: I) -> PyResult<PySet> where I: ToPyObject {
+        iterable.with_borrowed_ptr(py, |iterable| unsafe {
+            err::result_cast_from_owned_ptr(py, ffi::PySet_New(iterable))
+        })
+    }
+
+    /// Creates an empty set
+    ///
+    /// Corresponds to `set()` in Python
+    #[inline]
+        pub fn empty(py: Python) -> PyResult<PySet> {
+        unsafe {
+            err::result_cast_from_owned_ptr(py,
+                ffi::PySet_New(ptr::null_mut()))
+        }
+    }
+
+    /// Empty an existing set of all values.
+    #[inline]
+    pub fn clear(&self, py: Python) -> PyResult<()> {
+        unsafe { err::error_on_minusone(py,
+                     ffi::PySet_Clear(self.0.as_ptr())
+                 )}
+    }
+
+    /// Return the number of items in the set
+    /// This is equivalent to Python `len(self)`
+    #[inline]
+    pub fn len(&self, _py: Python) -> usize {
+        unsafe { ffi::PySet_Size(self.0.as_ptr()) as usize }
+    }
+
+    /// Determine if the set contains the specified value.
+    /// This is equivalent to the Python expression `value in self`.
+    pub fn contains<V>(&self, py: Python, value: V) -> PyResult<bool> where V: ToPyObject {
+        value.with_borrowed_ptr(py, |key| unsafe {
+            match ffi::PySet_Contains(self.0.as_ptr(), key) {
+                1 => Ok(true),
+                0 => Ok(false),
+                _ => Err(PyErr::fetch(py))
+            }
+        })
+    }
+
+    /// Add a value.
+    /// This is equivalent to the Python expression `self.add(value)`.
+    pub fn add<V>(&self, py: Python, value: V) -> PyResult<()> where V: ToPyObject {
+        value.with_borrowed_ptr(py, |value| unsafe {
+            err::error_on_minusone(py,
+                ffi::PySet_Add(self.0.as_ptr(), value))
+        })
+    }
+
+    /// Discard a value
+    /// This is equivalent to the Python expression `self.discard(value)`.
+    pub fn discard<V>(&self, py: Python, value: V) -> PyResult<()> where V: ToPyObject {
+        value.with_borrowed_ptr(py, |value| unsafe {
+            err::error_on_minusone(py,
+                ffi::PySet_Discard(self.0.as_ptr(), value))
+        })
+    }
+
+    /// Pop a value
+    /// This is equivalent to the Python expression `self.pop(value)`.
+    /// We get KeyError if the set is empty
+    pub fn pop(&self, py: Python) -> PyResult<PyObject> {
+        let as_opt = unsafe {
+            PyObject::from_borrowed_ptr_opt(py,
+                ffi::PySet_Pop(self.0.as_ptr()))
+        };
+        match as_opt {  // TODO use one of the magics
```

Comment:
I don't think you need to use one of the utilities for the error.  This code is clear enough.  Just drop the TODO.